### PR TITLE
Make Kestrel prefer on-disk cert for macOS bind.

### LINKED ISSUE #44754: Running ASP.NET Core apps using HTTPS on macOS results in Keychain Access prompts
It seems we regressed some of the work we did for .NET 7 around improving the developer experience on macOS, specifically the change to have Kestrel load the developer HTTPS cert (i.e. the one created by `dotnet dev-certs https`) from a .pfx file on disk rather than from the macOS Keychain in order to avoid Keychain Access prompts for every new ASP.NET Core app created that's run with the https launch profile.

The regression was made in PR #43358 with the code change meaning that the matching dev cert from the Keychain is always preferred over the dev cert loaded from the .pfx file.

We'll need to ensure that when the dev cert is being loaded for the purpose of a Kestrel HTTPS endpoint, that the file on disk is preferred. Other scenarios (e.g. operations performed by `dotnet dev-certs https`) will likely require the cert to be loaded from the Keychain (at least the scenario that involves exporting it from the Keychain to a .pfx file on disk will :smile:), but the details of that will need to be determined and confirmed.

### PR DESCRIPTION
**Pull request by adityamandaleeka**
Fix #44754

After the macOS dev certs overhaul, https://github.com/dotnet/aspnetcore/pull/43358 fixed an issue where certs on disk were being preferred in cases (like export) that should only use exportable certs (i.e. not the on-disk ones). However, the codepath updated by that PR is also taken during Kestrel cert binding, so I effectively un-did a useful part of the dev-cert UX improvements.

The aim of this PR is to fix that issue, preferring the on-disk cert for bind while also ensuring that cases that require exportability will get the keychain version of the cert.

### CODE REVIEW

**Comment by adityamandaleeka** [comment]
@javiercn PTAL

**Review by javiercn — APPROVED** [review]

File: src/Shared/CertificateGeneration/MacOSCertificateManager.cs [hunk 1]
```diff
@@ -373,25 +373,27 @@ protected override IList<X509Certificate2> GetCertificatesToRemove(StoreName sto
...
             // .NET 7+ installs the certificate on disk as well as on the user keychain (for backwards
             // compatibility with pre-.NET 7).
-            // Note that the actual certs we populate need to be the ones from the store location, and
-            // not the version from disk, since we may do other operations with these certs later (such
-            // as exporting) which would fail with crypto errors otherwise.
-            var onDiskAndKeychain = certsFromStore.Intersect(certsFromDisk, ThumbprintComparer.Instance);
+            // Note that if we require exportable certs, the actual certs we populate need to be the ones
+            // from the store location, and not the version from disk. If we don't require exportability,
+            // we favor the version of the cert that's on disk (avoiding unnecessary keychain access
+            // prompts).
```
Comment (halter73) [inline src/Shared/CertificateGeneration/MacOSCertificateManager.cs:394]:
```suggestion
            // prompts). Intersect returns items from the first collection.
```

Maybe this is a little bit too nuanced to rely on. I don't think the API docs describe this behavior, but I doubt linq would ever change this.

**Review by halter73 — APPROVED** [review]

File: src/Shared/CertificateGeneration/MacOSCertificateManager.cs (hunk 1 above)
Comment (adityamandaleeka) [reply src/Shared/CertificateGeneration/MacOSCertificateManager.cs:394]: It's described in the Remarks section of the API doc for Intersect:

> The intersection of two sets A and B is defined as the set that contains all the elements of A that also appear in B, but no other elements.

But sure, we can also be explicit here.

**Pull request merged**